packages/workbox-broadcast-cache-update/BroadcastCacheUpdate.mjs

```diff
@@ -32,32 +42,101 @@ class BroadcastCacheUpdate {
 
-    this._channelName = channelName;
-    this._headersToCheck = headersToCheck || [
-      'content-length',
-      'etag',
-      'last-modified',
-    ];
-    this._source = source || 'workbox-broadcast-cache-update';
+    // The message listener needs to be added in the initial run of the
+    // service worker, but since we don't actually need to be listening for
+    // messages until the cache updates, we only invoke the callback if set.
+    this._onReadyMessageCallback = null;
+    self.addEventListener('message', (event) => {
+      if (event.data.type === 'WINDOW_READY' &&
+          event.data.meta === 'workbox-window') {
+        if (this._onReadyMessageCallback) {
+          if (process.env.NODE_ENV !== 'production') {
+            logger.debug(`Received WINDOW_READY event: `, event);
+          }
+          this._onReadyMessageCallback();
+        }
+      }
+    });
+  }
+
+  /**
+   * Compare two [Responses](https://developer.mozilla.org/en-US/docs/Web/API/Response)
+   * and send a message via the
+   * {@link https://developers.google.com/web/updates/2016/09/broadcastchannel|Broadcast Channel API}
+   * if they differ.
+   *
+   * Neither of the Responses can be {@link http://stackoverflow.com/questions/39109789|opaque}.
+   *
+   * @param {Object} options
+   * @param {Response} options.oldResponse Cached response to compare.
+   * @param {Response} options.newResponse Possibly updated response to compare.
+   * @param {string} options.url The URL of the request.
+   * @param {string} options.cacheName Name of the cache the responses belong
+   *     to. This is included in the broadcast message.
+   * @param {Event} [options.event] event An optional event that triggered
+   *     this possible cache update.
+   * @return {Promise} Resolves once the update is sent.
+   */
+  notifyIfUpdated({oldResponse, newResponse, url, cacheName, event}) {
+    if (!responsesAreSame(oldResponse, newResponse, this._headersToCheck)) {
+      if (process.env.NODE_ENV !== 'production') {
+        logger.log(`Newer response found (and cached) for:`, url);
+      }
+
+      const sendUpdate = async () => {
+        // In the case of a navigation request, the requesting page will likely
+        // not have loaded its JavaScript in time to recevied the update
+        // notification, so we defer it until ready (or we timeout waiting).
+        if (event && event.request.mode === 'navigate') {
+          if (process.env.NODE_ENV !== 'production') {
+            logger.debug(`Original request was a navigation request, ` +
+                `waiting for a ready message from the window`, event.request);
+          }
+          await this._windowReadyOrTimeout();
+        }
+        await broadcastUpdate({channel: this._getChannel(), cacheName, url});
+      };
 
-    // TODO assert typeof headersToCheck instanceof Array
+      // Send the update and ensure the SW stays alive until it's sent.
+      const done = sendUpdate();
+      if (event) {
```

Comment:
I can, but I don't think this code is executed async. (Is `cacheDidUpdate` ever executed async?)

Reply:
Yeah, `cacheDidUpdate`s are executed outside of synchronous `fetch` handler execution since they always follow a call to `await cache.put(...)` (or the wrapped equivalent).

Reply:
Sure -- but for a future discussion: I wonder how much code we'd save by switching to use Jake's polyfill.

Reply:
One of my concerns was that [Jake's code](https://github.com/jakearchibald/async-waituntil-polyfill/blob/master/async-waituntil.js) can't actually check to see if there's support or not for async `waitUntil()`, and it unconditionally overwrites both `ExtendableEvent.prototype.waitUntil` and `FetchEvent.prototype.respondWith` with the polyfill code.

Running that code in all browsers, even those with native supports, seems like something worth avoiding?

Reply:
I understand the concern, but our implementation is also adding extra code that needs to run in all browsers (even those with native support).

My concern is more that we're potentially not working in some cases (in some browsers) in subtle ways that we don't understand because they only appear in the wild. And if users are only testing on Chrome, they're not seeing those errors either (this may be unfounded, I'm not sure).

I wonder if there's a middle ground where we don't polyfill the behavior for all events, just the ones where we *know* we're going to be calling `event.waitUntil()` async ourselves.

Anyway, like I said, I think it's a separate discussion.